File: cli/cjtCli/internal/daemon/proxy/forwarder.go
package proxy

import (
	"bytes"
	"cjtCli/internal/core/telemetry"
	"cjtCli/internal/daemon/dlq"
	"fmt"
	"math"
	"net/http"
	"time"

	"com.chanjet/connector-sdk-go/pkg/protocol"
	"go.uber.org/zap"
)

// Forwarder handles forwarding events to local webhook targets
type Forwarder interface {
	Forward(event protocol.EventFrame, targetURL string) error
}

type simpleForwarder struct {
	tel    *telemetry.Telemetry
	dlq    dlq.Store
	client *http.Client
}

func NewForwarder(tel *telemetry.Telemetry, dlq dlq.Store) Forwarder {
	return &simpleForwarder{
		tel:    tel,
		dlq:    dlq,
		client: &http.Client{Timeout: 5 * time.Second},
	}
}

func (f *simpleForwarder) Forward(event protocol.EventFrame, targetURL string) error {
	if targetURL == "" {
		return fmt.Errorf("target URL is empty")
	}

	go f.forwardWithRetry(event, targetURL)
	return nil
}

func (f *simpleForwarder) forwardWithRetry(event protocol.EventFrame, targetURL string) {
	maxRetries := 5
	var lastErr error

	for i := 0; i <= maxRetries; i++ {
		if i > 0 {
			delay := time.Duration(math.Pow(2, float64(i-1))) * time.Second
			f.tel.Sys().Warn("Retrying event forward", 
				zap.String("msg_id", event.MsgID), 
				zap.Int("attempt", i), 
				zap.Duration("delay", delay))
			time.Sleep(delay)
		}

		err := f.doForward(event, targetURL)
		if err == nil {
			f.tel.Audit().Info("Event delivered successfully", zap.String("msg_id", event.MsgID))
			return
		}

		lastErr = err
		f.tel.Sys().Warn("Forward attempt failed", zap.String("msg_id", event.MsgID), zap.Error(err))
	}

	// All retries failed, sink to DLQ
	f.tel.DLQ().Error("Event delivery failed, sinking to DLQ", zap.String("msg_id", event.MsgID), zap.Error(lastErr))
	if err := f.dlq.Save(event, lastErr.Error()); err != nil {
		f.tel.Sys().Error("CRITICAL: Failed to save to DLQ", zap.Error(err))
	}
}

func (f *simpleForwarder) doForward(event protocol.EventFrame, targetURL string) error {
	req, err := http.NewRequest("POST", targetURL, bytes.NewBuffer([]byte(event.Payload)))
	if err != nil {
		return err
	}

	// Pass headers
	for k, v := range event.Headers {
		req.Header.Set(k, v)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-MSG-ID", event.MsgID)

	resp, err := f.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("target returned status %d", resp.StatusCode)
	}

	return nil
}

File: cli/cowen/src/daemon/forwarder.rs
use reqwest::Client;
use serde_json::Value;
use std::time::Duration;
use crate::daemon::dlq::DlqStore;
use std::sync::Arc;

#[derive(Clone)]
pub struct Forwarder {
    client: Client,
    dlq: Arc<DlqStore>,
    target_url: String,
}

impl Forwarder {
    pub fn new(dlq: Arc<DlqStore>, target_url: &str) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(10))
            .build()
            .unwrap_or_else(|_| Client::new());

        Self {
            client,
            dlq,
            target_url: target_url.to_string(),
        }
    }

    pub async fn forward(&self, event: Value) {
        if self.target_url.is_empty() {
            tracing::warn!(target: "stream", "No webhook_target configured. Event dropped locally.");
            println!("⚠️ No webhook_target configured. Event dropped locally.");
            return;
        }

        let msg_id = event.get("id").and_then(|v| v.as_str()).unwrap_or("unknown_id").to_string();
        let msg_type = event.get("msgType").and_then(|v| v.as_str()).unwrap_or("UNKNOWN").to_string();
        let headers = event.get("headers").map(|v| v.to_string()).unwrap_or_else(|| "{}".to_string());
        let payload = serde_json::to_string(&event).unwrap_or_else(|_| "{}".to_string());

        tracing::info!(target: "stream", msg_id = %msg_id, msg_type = %msg_type, target = %self.target_url, "Forwarding event to webhook");
        println!("➡️ Forwarding event [{}] to {}...", msg_type, self.target_url);

        let resp = self.client.post(&self.target_url)
            .header("Content-Type", "application/json")
            .body(payload.clone())
            .send()
            .await;

        match resp {
            Ok(r) if r.status().is_success() => {
                tracing::info!(target: "stream", msg_id = %msg_id, status = %r.status(), "Event successfully forwarded");
                println!("✅ Successfully forwarded event [{}]", msg_id);
            }
            Ok(r) => {
                let err_msg = format!("HTTP error: {}", r.status());
                tracing::error!(target: "stream", msg_id = %msg_id, status = %r.status(), "Forward failed, saving to DLQ");
                println!("❌ Forward failed: {}", err_msg);
                let _ = self.dlq.save(&msg_id, &msg_type, &payload, &headers, &err_msg);
            }
            Err(e) => {
                let err_msg = format!("Network error: {}", e);
                tracing::error!(target: "stream", msg_id = %msg_id, error = %e, "Forward network failed, saving to DLQ");
                println!("❌ Forward network failed: {}", err_msg);
                let _ = self.dlq.save(&msg_id, &msg_type, &payload, &headers, &err_msg);
            }
        }
    }
}
